crate/interfaces/src/hsm/hsm_store.rs

```
@@ -236,9 +237,178 @@ impl ObjectsStore for HsmStore {
+    if let Some(s) = state {
+        if s != State::Active {
+            return false;
+        }
+    }
+    if researched_attributes.link.is_some() {
+        return false;
+    }
+    if !researched_attributes.get_tags().is_empty() {
+        return false;
+    }
+    if researched_attributes.object_group.is_some() {
+        return false;
+    }
+    if researched_attributes.object_group_member.is_some() {
+        return false;
+    }
+    if researched_attributes.comment.is_some() {
+        return false;
+    }
+    if researched_attributes.contact_information.is_some() {
+        return false;
+    }
+    if let Some(critical) = researched_attributes.critical {
+        if critical {
+            return false;
+        }
+    }
+    if researched_attributes.description.is_some() {
+        return false;
     }
+    if researched_attributes.digest.is_some() {
+        return false;
+    }
+    if researched_attributes.short_unique_identifier.is_some() {
+        return false;
+    }
+    if researched_attributes.cryptographic_usage_mask.is_some() {
+        return false;
+    }
+    if researched_attributes.x_509_certificate_identifier.is_some() {
+        return false;
+    }
+    if researched_attributes.x_509_certificate_issuer.is_some() {
+        return false;
+    }
+    if researched_attributes.x_509_certificate_subject.is_some() {
+        return false;
+    }
+    true
+}
+
+/// Derives the HSM object filter from attributes.
+/// Returns `None` if incompatible combination.
+fn derive_hsm_object_filter(researched_attributes: &Attributes) -> Option<HsmObjectFilter> {
+    let mut object_filter = HsmObjectFilter::Any;
+
+    if let Some(cryptographic_algorithm) = researched_attributes.cryptographic_algorithm {
+        object_filter = match cryptographic_algorithm {
+            CryptographicAlgorithm::AES => HsmObjectFilter::AesKey,
+            CryptographicAlgorithm::RSA => HsmObjectFilter::RsaKey,
+            _ => return None,
```

If you don't mind, I would return an error instead of None

---

and same for other filters please